pkg/infra/game/stage/loot/loot.go

```diff
@@ -24,8 +25,50 @@ func UpdateItems(state state.State, agents map[commons.ID]agent.Agent) state.Sta
 	return updatedState
 }
 
-func AllocateLoot(globalState state.State, weaponLoot []uint, shieldLoot []uint) state.State {
+// HPi and STi are the index of HP potion slice and ST potion slice that is allocate to the agent. Pass one at a time.
+func AllocateHPPotion(globalState state.State, agentID commons.ID, HPi int) (state.State, immutable.Map[int, uint]) {
 	allocatedState := globalState
+	hpPotionValue := allocatedState.PotionSlice.HPpotion[HPi]
+	v := allocatedState
+	a := allocatedState.AgentState[agentID]
+	a.Hp = v.AgentState[agentID].Hp + hpPotionValue
+	allocatedState.AgentState[agentID] = a
+	allocatedState.PotionSlice.HPpotion[HPi] = 0
+	hpPotionMap := commons.Slice2Map(allocatedState.PotionSlice.HPpotion)
+	HPPotionMap := commons.MapToImmutable(hpPotionMap)
+	return allocatedState, HPPotionMap
```

Comment:
That was before, I was thinking storing a simple slice in the global state and generating a temporary immu list for agents to view, but that was before I figure out how the view.go is expected to work. That's why I put a _ to the immu list output, I didn't know where to use these immu lists from outputs.
I realised it's probably still better to use an immutable potion list in the global state, and every time when making an allocation change, do (immu list -> slice -> immu list) and make modifications on the slice. and that slice is temporary.

Reply:
Or keep it the old way, but wrap the potion slice to immutable list in the view.go?

Reply:
Ah I see. I think keeping it as a slice/array in globalstate is fine because we 'trust' the infra code. It only needs to be turned to immutible when passed to the agents (through view for example).